File: Questão4.go
package main

import "fmt"


type Animal interface {
    Som() string
}


type Cachorro struct{}


func (Cachorro) Som() string {
    return "Au Au"
}


type Gato struct{}


func (Gato) Som() string {
    return "Miau"
}

func emitirSom(animal Animal) {
    fmt.Println(animal.Som())
}

func main() {
    cachorro := Cachorro{}
    gato := Gato{}

    emitirSom(cachorro) 
    emitirSom(gato)     
}

File: Questão4.py
class Animal:
    """Classe base que representa um animal."""

    def som(self) -> str:
        raise NotImplementedError("Subclasse deve implementar este método.")

class Cachorro(Animal):
    def som(self) -> str:
        return "Au Au"

class Gato(Animal):
    def som(self) -> str:
        return "Miau"


def emitir_som(animal: Animal) -> None:
    print(animal.som())

cachorro = Cachorro()
gato = Gato()

emitir_som(cachorro)  
emitir_som(gato)      
